### DISCUSSION TITLE
Closing Sheet component from within a component inside the SheetContent

### DISCUSSION THREAD

**Discussion by AbhayaShankar** [shadcn-ui/ui#1790] [Q&A] [answered]
I have a use case where I wanna close the sheet component when I click on any link.
But the thing is the all links are wrapped inside another component.

```tsx
 <Sheet>
      <SheetTrigger asChild>
        <Button variant={"ghost"} size={"icon"} className="lg:hidden">
          <Menu />
        </Button>
      </SheetTrigger>
      <SheetContent side={"left"} className="p-0">
        <Navbar />
      </SheetContent>
 </Sheet>
``` 

```tsx
{NavbarLinks.map((item) => (
          <Link
            key={item.id}
            className={cn(
              "font-semibold transition-all delay-75 text-[17px] dark:text-muted-foreground 
               dark:hover:text-white  text-gray-500 hover:text-gray-900 ",
              font.className
            )}
            href={item.href}
          >
            {item.label}
          </Link>
))}
```

And, I want the `SheetClose` component to close the sheet when a link which is inside `Navbar` Component is clicked. How can it be done ?

![image](https://github.com/shadcn-ui/ui/assets/113185442/845cf8ac-df0e-47e8-804a-15e4da5f7881)

Upon clicking on these links the sheet component closes. Can someone suggest me a way of doing that!

**Comment by joaom00** [comment]
You can do:

```tsx
{NavbarLinks.map((item) => (
     <SheetClose asChild>
          <Link
            key={item.id}
            className={cn(
              "font-semibold transition-all delay-75 text-[17px] dark:text-muted-foreground 
               dark:hover:text-white  text-gray-500 hover:text-gray-900 ",
              font.className
            )}
            href={item.href}
          >
            {item.label}
          </Link>
    </SheetClose>
))}
```

**Reply by lucasgustavodiaz** [reply]
i have the same doubt, Is this the most optimal way to do it? thx

**Reply by AbhayaShankar** [reply]
But @joaom00 This ain't working.
It throws an error while working in Next13.
Error is 

```bash
Unhandled Runtime Error
Error: `DialogClose` must be used within `Dialog`
```

Let me emphasis once again where I am using this-->

`Sidebar.tsx`

```tsx
 <Sheet>
      <SheetTrigger asChild>
        <Button variant={"ghost"} size={"icon"} className="lg:hidden">
          <Menu />
        </Button>
      </SheetTrigger>
      <SheetContent side={"left"} className="p-0">
        <Navbar />
      </SheetContent>
 </Sheet>
```

And the links lie inside the Navbar component. below is the Navbar component where I am iterating over a array and correspondingly rendering them.

`Navbar.tsx`

```tsx
export const Navbar = () => {
  return (
    <nav>
        <div >
        {NavbarLinks.map((item) => (
          <SheetClose asChild key={item.id}>
            <Link
              key={item.id}
              className={cn(
                "font-semibold transition-all delay-75 text-[17px] dark:text-muted-foreground dark:hover:text-white  text-gray-500 hover:text-gray-900 ",
                font.className
              )}
              href={item.href}
            >
              {item.label}
            </Link>
          </SheetClose>
        ))}
      </div>
      <div className="block lg:hidden absolute right-5">
        <ModeToggle />
      </div>
      <div className="hidden lg:block">
        <ModeToggle />
      </div>
    </nav>
  );
};
```

![image](https://github.com/shadcn-ui/ui/assets/113185442/387cb33c-1b89-474f-8d8d-5a9f70c0a67e)

Upon wrapping <SheetClose > ... </SheetClose> , I am getting error. How do I make changes to proceed ?

**Reply by joaom00** [reply]
Are you using `<Navbar />` somewhere outside the `<Sheet/>`?

**Reply by AbhayaShankar** [reply]
Yes. It's a separate component.

**Reply by joaom00** [reply]
Yeah, so you need to render `<SheetClose />` only when `<Navbar />` in inside `<SheetContent />`

**Reply by joaom00** [reply]
> i have the same doubt, Is this the most optimal way to do it? thx

I would say so. another way is to control the open state with `open` and `onOpenChange` props and pass down the setter state

**Reply by AbhayaShankar** [reply]
So let's say I pass on the `open` state and `onOpenChange` props such like this

```tsx
// sheetOpen state
const [sheetOpen, setSheetOpen] = useState(false);

 <Sheet  open={sheetOpen} onOpenChange={setSheetOpen}>
      {...content}
      <SheetContent side={"left"} className="p-0">
        <Navbar />
      </SheetContent>
 </Sheet>
```
How do I pass on the `props` down to `Navbar` component 
How does the Sheet component knows that the `SheetClose` is being called from inside the `SheetContent`?

```tsx
 {NavbarLinks.map((item) => (
          <SheetClose asChild key={item.id}>
            <Link
              key={item.id}
              href={item.href}
            >
              {item.label}
            </Link>
          </SheetClose>
        ))}
```

Can you suggest a way of achieving this ?

**Reply by joaom00** [reply, marked as answer]
Here an example: https://codesandbox.io/s/upbeat-galileo-6rxfm7?file=/src/App.tsx

> How does the Sheet component knows that the SheetClose is being called from inside the SheetContent?

because it uses context to close the sheet so it needs to be inside `<Sheet />` which is the context provider

**Reply by AbhayaShankar** [reply]
Great! Thanks for the explanation @joaom00  That helps a lot. 🔥

**Reply by HamedMP** [reply]
Thanks for sharing both solutions, `<SheetClose>` is way more performant (on my dev) than using States. With states the page loads on each sheet open or link click.

**Reply by sahnisarbjit** [reply]
Thanks @joaom00. Adding more, if you use SSR (For example Next.js server component), then `<SheetClose>` is the way to go. In other solution, navigation links won't be visible in HTML response. Hence bots can't see links to easily navigate the site.

**Reply by DiegoFelipe** [reply]
Thank You @joaom00

**Comment by Hombre2014** [comment]
Thank you very much @joaom00 !

**Comment by itpaulin** [comment]
Thank you @joaom00

**Comment by PritamMagdum** [comment]
Thank You @joaom00

**Comment by jadejamig** [comment]
You saved me @joaom00

**Comment by MANOJ-M-01** [comment]
Thank you @joaom00

**Comment by michaelcanonizado** [comment]
Thanks @joaom00🙌!

Went with the second option. But it seems to introduce another issue when you are rendering the <Navbar/> component within another section of the website that doesn't need to trigger the sheet close and the setOpen() function. It totally hard reloads the page and throws a NextJS error.

Solved this by marking setOpen() as optional within the <Navbar/> component:

```
const Navbar = ({
setOpen
}:{
setOpen?: (value: boolean) => void;
}) => {
...
}
```

and conditional setting it on the onClick prop:

```
<Link
    key={item.id}
    href={item.href}
    onClick={setOpen ? () => setOpen(false) : () => {}}
/>
```

**Comment by Ahmad-Omar-design** [comment]
Thank you so much @joaom00 !!
